libenv/unix_iface.c

```diff
@@ -197,33 +196,59 @@ static void GetMacAddress(EvalContext *ctx, ARG_UNUSED int fd, struct ifreq *ifr
     {
         if ( strcmp(ifa->ifa_name, ifp->ifr_name) == 0)
         {
+            if (ifa->ifa_addr == NULL)
+            {
+               Log(LOG_LEVEL_VERBOSE, "Interface '%s' has no address information", ifa->ifa_name);
+               continue;
+            }
+#if AF_LINK
             if (ifa->ifa_addr->sa_family == AF_LINK)
             {
-                sdl = (struct sockaddr_dl *)ifa->ifa_addr;
-                m = (char *) LLADDR(sdl);
-
-                snprintf(hw_mac, sizeof(hw_mac), "%.2x:%.2x:%.2x:%.2x:%.2x:%.2x",
-                    (unsigned char) m[0],
-                    (unsigned char) m[1],
-                    (unsigned char) m[2],
-                    (unsigned char) m[3],
-                    (unsigned char) m[4],
-                    (unsigned char) m[5]);
-
-                EvalContextVariablePutSpecial(ctx, SPECIAL_SCOPE_SYS, name, hw_mac, CF_DATA_TYPE_STRING, "source=agent");
-                if (!RlistContainsString(*hardware, hw_mac))
-                {
-                    RlistAppend(hardware, hw_mac, RVAL_TYPE_SCALAR);
-                }
-                RlistAppend(interfaces, ifa->ifa_name, RVAL_TYPE_SCALAR);
+                struct sockaddr_dl *sdl = (struct sockaddr_dl *)ifa->ifa_addr;
+                mac_pointer = (char *) LLADDR(sdl);
+            }
+#elif AF_PACKET
+            if (ifa->ifa_addr->sa_family == AF_PACKET)
+            {
+                struct sockaddr_ll sll = (struct sockaddr_ll *)ifa->ifa_addr;
+                mac_allocd = xmalloc(sll->sll_halen);
+                memcpy(mac_allocd, sll->sll_addr, sll->sll_halen);
```

Comment:
Why do we need the copy? Wouldn't it be possible to just access `sll->sll_addr` directly?

Reply:
Indeed. Re-working that.